hudi-client/hudi-flink-client/src/test/java/org/apache/hudi/merge/TestHoodieFlinkRecordMerger.java

```diff
@@ -0,0 +1,180 @@
+import org.apache.hudi.common.model.HoodieEmptyRecord;
+import org.apache.hudi.common.model.HoodieKey;
+import org.apache.hudi.common.model.HoodieOperation;
+import org.apache.hudi.common.model.HoodieRecord;
+import org.apache.hudi.common.testutils.HoodieTestDataGenerator;
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.collection.Pair;
+import org.apache.hudi.util.AvroSchemaConverter;
+
+import org.apache.avro.Schema;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.logical.RowType;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.apache.hudi.common.testutils.HoodieTestDataGenerator.TRIP_EXAMPLE_SCHEMA;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Tests for {@code HoodieFlinkRecordMerger}.
+ */
+public class TestHoodieFlinkRecordMerger {
+
+  private static final RowType RECORD_ROWTYPE = new RowType(
+      false,
+      Arrays.asList(
+          new RowType.RowField(HoodieRecord.COMMIT_TIME_METADATA_FIELD, DataTypes.STRING().getLogicalType()),
+          new RowType.RowField(HoodieRecord.COMMIT_SEQNO_METADATA_FIELD, DataTypes.STRING().getLogicalType()),
+          new RowType.RowField(HoodieRecord.RECORD_KEY_METADATA_FIELD, DataTypes.STRING().getLogicalType()),
+          new RowType.RowField(HoodieRecord.PARTITION_PATH_METADATA_FIELD, DataTypes.STRING().getLogicalType()),
+          new RowType.RowField(HoodieRecord.FILENAME_METADATA_FIELD, DataTypes.STRING().getLogicalType()),
+          new RowType.RowField("uuid", DataTypes.STRING().getLogicalType()),
+          new RowType.RowField("partition", DataTypes.STRING().getLogicalType()),
+          new RowType.RowField("f1_int", DataTypes.INT().getLogicalType()),
+          new RowType.RowField("f2_string", DataTypes.STRING().getLogicalType()),
+          new RowType.RowField("ts", DataTypes.BIGINT().getLogicalType())));
+
+  private static final String RECORD_KEY = "key";
+  private static final String PARTITION = "partition";
+
+  /**
+   * If the input records are not Flink HoodieRecord, it throws.
+   */
+  @Test
+  void testMergerWithAvroRecord() {
+    try (HoodieTestDataGenerator dataGenerator = new HoodieTestDataGenerator(0L)) {
+      List<HoodieRecord> records = dataGenerator.generateInserts("001", 2);
+      EventTimeFlinkRecordMerger merger = new EventTimeFlinkRecordMerger();
+      TypedProperties props = new TypedProperties();
+      Schema recordSchema = new Schema.Parser().parse(TRIP_EXAMPLE_SCHEMA);
+      assertThrows(
+          IllegalArgumentException.class,
+          () -> merger.merge(records.get(0), recordSchema, records.get(1), recordSchema, props));
```

Comment:
would the check be removed in final state? we should ensure the record type from the invokers instead of such per-row redundant check.

Reply:
The check is a comparition of enum variables`older.getRecordType() == HoodieRecord.HoodieRecordType.FLINK`, and the cost is negligible. Anyway, it would be better if we have chance to ensure the type from high level, I'll add some comments here as reminder.